man/unibdf2hex.1
.TH UNIBDF2HEX 1 "2013 Jul 07"
.SH NAME
unibdf2hex \- Convert BDF font glyphs into Unifont .hex glyphs
.SH SYNOPSIS
.br
.B unibdf2hex
<
.I input\-font.bdf
>
.I output\-font.hex
.SH DESCRIPTION
.B unibdf2hex
reads a BDF font file, extracting selected code points and printing
them on stdout in Unifont .hex format.  This program was used to
extract CJK ideographs from the 16x16 version of Wen Quan Yi's BDF
font.  Currently the program is hard\-coded to only extract those
code points that comprise the "wqy.hex" source font file used to
build "unifont.hex" but this source code could easily be modified.
.SH FILES
*.hex GNU Unifont font files
.SH SEE ALSO
.BR bdfimplode(1),
.BR hex2bdf(1),
.BR hex2sfd(1),
.BR hexbraille(1),
.BR hexdraw(1),
.BR hexkinya(1),
.BR hexmerge(1),
.BR johab2ucs2(1),
.BR unibmp2hex(1),
.BR unicoverage(1),
.BR unidup(1),
.BR unifont(5),
.BR unifont\-viewer(1),
.BR unifontchojung(1),
.BR unifontksx(1),
.BR unifontpic(1),
.BR unigencircles(1),
.BR unigenwidth(1),
.BR unihex2bmp(1),
.BR unihex2png(1),
.BR unihexfill(1),
.BR unihexgen(1),
.BR unipagecount(1),
.BR unipng2hex(1)
.SH AUTHOR
.B unibdf2hex
was written by Paul Hardy.
.SH LICENSE
.B unibdf2hex
is Copyright \(co 2008 Paul Hardy.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH BUGS
No known bugs exist.


NEWS
2015-07-02: Release 9.0.01

   Updates to support Unicode 9.0.0.

2015-06-28: Release 8.0.01

  Updates to support Unicode 8.0.0.

2014-10-23: Release 7.0.06

   A few new scripts were added:
      U+010350..U+01037F Old Permic
      U+01F650..U+01F67F Ornamental Dingbats
      U+01F780..U+01F7FF Geometrical Shapes Extended
      U+01F800..U+01F8FF Supplemental Arrows-C

   Various changes were made to Makefiles, Perl scripts, and one bash
   shell script to conform to Debian Policy Manual version 3.9.6.

   Several changes were made to Makefiles:
      * Use "set -e" at the start of multi-command shell invocations,
        to trap errors.
      * Use "&&" rather than ";" in multi-command shell invocations,
        to fail on errors.
      * install now uses "-s" flag for C program binaries.

   Other minor changes were made to scripts; see ChangeLog
   for details.

2014-10-17: Release 7.0.05

   Modified ASCII letter 'l' so it would have the same height as
   the other lower-case letters with ascenders; this is one pixel
   taller than the capital letters.  Lowered the horizontal stroke 
   in 't' by one pixel so it would line up with the horizontal
   stroke in 'f'; this is one pixel lower than the x-height of
   the ASCII letters.  Letters elsewhere in Plane 0 that were formed
   with 'k', 'l', or 't' were checked and modified where appropriate
   to conform with the new ASCII glyphs.

   Completed the Miscellaneous Symbols and Pictographs script in Plane 1,
   and modified Plane 1 Emoticons and Transport and Map Symbols scripts.

   Now create file font/compiled/plane01.html, showing a color-coded
   indication of completion of all of Unicode Plane 1 as an HTML
   table in a self-contained HTML file.  That and the compiled/index.html
   are the two HTML files that building the font creates.  The plane01.html
   file gives a color coding to show completion of each 256-glyph block,
   from 0% (pure red) to 100% (pure green).

2014-10-11: Release 7.0.04

   Added five more Supplemental Multilingual Plane (Plane 1) scripts
   and improved others thanks to several contributors.  There are now
   over 5,000 drawn assigned glyphs for Plane 1!

   Reorganized font/plane01 directory.  Several other changes; see
   ChangeLog file for details.

   unihex2sfd and unipagecount now handle all Unicode planes, not just
   Plane 0.


2014-07-01: Release 7.0.03

   Minor release for Debian.  Added a man page for unihexfill(1);
   Debian requires a man page for each program.

2014-06-22: Release 7.0.02

   Fixes a bug in font/Makefile: there were two "done" statements at the
   end of one "for" loop.  Also adds many new Supplemental Multilingual
   Plane (Plane 1) scripts.  See the ChangeLog file for details.

2014-06-21: Release 7.0.01

   Provides a glyph for every visible code point in the Unicode 7.0 Basic
   Mulitlingual Plane.  Also adds more scripts in the Unicode 7.0
   Supplemental Multilingual Plane.  See the ChangeLog file for full
   details.

   Note that the versioning scheme has changed.  Instead of a date, this
   package will use an incrementing minor version number in the future
   ("01", "02", etc.).

2014-02-14: Release 6.3.20140214

   This version adds the GNU Free Documentation License 1.3 to the "COPYING"
   file.

2014-02-04: Release 6.3.20140204

   This version added Qianqian Fang's name to the COPYRIGHT string now
   passed to Fontforge when the TTF fonts are built, for his contributions
   contained in the file "font/plane00/wqy.hex".

   This version also removes "<img>" links to icons on external websites from
   the "hangul/hangul-generation.html" web page, which were retrieved and
   stored locally when the original web page was saved in the "hangul/"
   directory.

2014-02-02: Release 6.3.20140202

   This version is the first designed to easily build multiple fonts.
   To accomplish this, font/Makefile and font/ttfsrc/Makefile were
   modified to pass FONT and COMBINING parameters, giving the names
   of the font file and its associated combining.txt file.

   This version introduces a subset of Unicode's Supplemental Multilingual
   Plane (SMP, or Plane 1).  The file "font/plane01/plane01.hex" contains
   a glyph for every visible code point in the Unicode 5.0 SMP except
   "Cuneiform" and "Cuneiform Numbers and Punctuation"; those two scripts
   require more resolution than what will fit within a 16 by 16 pixel grid.
   Andrew Miller drew well over 1,000 glyphs for this effort, and added
   glyphs for several SMP scripts introduced after Unicode 5.0.

   This version also introduces a subset of the ConScript Unicode
   Registry (CSUR) Private Use Area glyphs.  Wojciech Stryjewski created
   glyphs for Tengwar and Klingon -- Tolkien fans and Trekkies rejoice!
   Paul Hardy drew Cirth to finish the Tolkien set, and drew other scripts.
   Andrew Miller made very extensive additions for the CSUR; see the
   ChangeLog for details.

   Finally, this version includes a special PSF font for using APL
   in console mode on GNU/Linux.  This PSF font contains 512 glyphs
   as a subset of Unifont; it will not work on systems that only
   support PSF fonts with 256 glyphs.  This PSF font will support
   GNU APL and several other APL implementations.

   The names of the font files are as follows:

      unifont-*               Font with Unifont Plane 0 glyphs

      unifont_sample-*        Font with Unifont glyphs, with combining
                              circles added for illustration

      unifont_csur-*          Font with glyphs from Unifont plus the
                              CSUR PUA glyphs

      unifont_upper-*         Font with glyphs from Planes 1 through 14

      unifont_upper_csur-*    Font with glyphs from Planes 1 through 15

      Unifont-APL8x16.psf.gz  The APL console font

Glyphs for code points U+FFFE and U+FFFF, previously added for illustration
purposes in unifont_sample, were removed because they caused problems on

src/unibdf2hex.c
/*
   unibdf2hex - program to convert a BDF file into a unifont.hex file.

   Author: Paul Hardy, January 2008

   Note: currently this has hard-coded code points for glyphs extracted
   from Wen Quan Yi to create the Unifont source file "wqy.hex".

   Copyright (C) 2008, 2013 Paul Hardy

   LICENSE:

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 2 of the License, or
      (at your option) any later version.
   
      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
      GNU General Public License for more details.
   
      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define UNISTART 0x3400
#define UNISTOP 0x4DBF

#define MAXBUF 256


int main() {
   int i;
// int j;
   int digitsout;  /* how many hex digits we output in a bitmap */
   int thispoint;
   char inbuf[MAXBUF];
   int bbxx, bbxy, bbxxoff, bbxyoff;

   int descent=4; /* font descent wrt baseline */
   int startrow;  /* row to start glyph        */
   unsigned rowout;

   while (fgets(inbuf, MAXBUF - 1, stdin) != NULL) {
      if (strncmp(inbuf, "ENCODING ", 9) == 0) {
         sscanf(&inbuf[9], "%d", &thispoint); /* get code point */
         /*
            If we want this code point, get the BBX (bounding box) and
            BITMAP information.
         */
         if ((thispoint >= 0x2E80 && thispoint <= 0x2EFF) || // CJK Radicals Supplement
             (thispoint >= 0x2F00 && thispoint <= 0x2FDF) || // Kangxi Radicals
             (thispoint >= 0x2FF0 && thispoint <= 0x2FFF) || // Ideographic Description Characters
             (thispoint >= 0x3001 && thispoint <= 0x303F) || // CJK Symbols and Punctuation (U+3000 is a space)
             (thispoint >= 0x3100 && thispoint <= 0x312F) || // Bopomofo
             (thispoint >= 0x31A0 && thispoint <= 0x31BF) || // Bopomofo extend
             (thispoint >= 0x31C0 && thispoint <= 0x31EF) || // CJK Strokes
             (thispoint >= 0x3400 && thispoint <= 0x4DBF) || // CJK Unified Ideographs Extension A
             (thispoint >= 0x4E00 && thispoint <= 0x9FCF) || // CJK Unified Ideographs
             (thispoint >= 0xF900 && thispoint <= 0xFAFF))   // CJK Compatibility Ideographs
            {
            while (fgets(inbuf, MAXBUF - 1, stdin) != NULL &&
                   strncmp(inbuf, "BBX ", 4) != 0); /* find bounding box */

            sscanf(&inbuf[4], "%d %d %d %d", &bbxx, &bbxy, &bbxxoff, &bbxyoff);
            while (fgets(inbuf, MAXBUF - 1, stdin) != NULL &&
                   strncmp(inbuf, "BITMAP", 6) != 0); /* find bitmap start */
            fprintf(stdout, "%04X:", thispoint);
//          j = 5; /* start after colon in .hex file line */
            digitsout = 0;
            /* Print initial blank rows */
            startrow = descent + bbxyoff + bbxy;

            /* Force everything to 16 pixels wide */
            for (i = 16; i > startrow; i--) {
               fprintf(stdout,"0000");
               digitsout += 4;
            }
            while (fgets(inbuf, MAXBUF - 1, stdin) != NULL &&
                   strncmp(inbuf, "END", 3) != 0) { /* copy bitmap until END */
               sscanf(inbuf, "%X", &rowout);
               /* Now force glyph to a 16x16 grid even if they'd fit in 8x16 */
               if (bbxx <= 8) rowout <<= 8;  /* shift left for 16x16 glyph */
               rowout >>= bbxxoff;
               fprintf(stdout, "%04X", rowout);
               digitsout += 4;
            }

            /* Pad for 16x16 glyph */
            while (digitsout < 64) {
               fprintf(stdout,"0000");
               digitsout += 4;
            }
            fprintf(stdout,"\n");
         }
      }
   }
   exit(0);
}
